nautobot/core/api/renderers.py

```
@@ -24,3 +36,168 @@ class NautobotJSONRenderer(JSONRenderer):
+
+        headers = self.get_headers(serializer_class(objects[0], context={"request": renderer_context["request"]}))
+
+        buffer = StringIO()
+        writer = csv.writer(buffer)
+        writer.writerow(headers)
+        for obj in objects:
+            writer.writerow(
+                self.object_to_row_elements(
+                    obj,
+                    headers=headers,
+                    serializer=serializer_class(obj, context={"request": renderer_context["request"]}),
+                )
+            )
+
+        return buffer.getvalue()
+
+    @classmethod
+    def get_headers(cls, serializer):
+        """Identify the appropriate CSV headers corresponding to the given serializer."""
+        base_headers = list(serializer.fields)
+
+        # Remove specific headers that we know are irrelevant
+        for undesired_header in [
+            "computed_fields",
+            "config_context",
+            "custom_fields",  # will be handled later as a special case
+            "local_config_context_data",
+            "notes_url",
+            "relationships",
+            "tree_depth",
+            "url",
+        ]:
+            if undesired_header in base_headers:
+                base_headers.remove(undesired_header)
+
+        # Remove all write-only and object-count headers
+        for header in list(base_headers):
+            if serializer.fields[header].write_only:
+                base_headers.remove(header)
+            if header.endswith("_count"):
+                base_headers.remove(header)
+
+        # Add individual headers for each relevant custom field
+        if "custom_fields" in serializer.fields:
+            cf_keys = CustomField.objects.get_for_model(serializer.Meta.model).values_list("key", flat=True)
+            cf_headers = sorted([f"cf_{key}" for key in cf_keys])
+        else:
+            cf_headers = []
+
+        # TODO: relationships? computed fields?
+
+        headers = base_headers + cf_headers
```

How do we plan to import Custom Fields if it doesn't exist on import (or the name has changed?) I figure this is going to have to be addressed in #3715

---

Given that the REST API normally silently ignores/discards any JSON keys that aren't understood and writable, I think it makes sense to do the same for custom fields that don't actually exist. In fact that might "just work".

---

Addressed - `cf_foobar` will be ignored if no `foobar` custom field exists, similar to how the REST API ignores any other fields/columns that it doesn't understand or are read-only.